### DISCUSSION TITLE
Vertexes and Edges handling

### DISCUSSION THREAD

**Discussion by f-hoenix** [ManevilleF/hexx#153] [Ideas]
Hello, thanks for this great crate 👍 
I'm writting a small game using it and I find myself in need of interacting with the edges between the Hex
I did not find a lot of code related to it in the repo, did I miss something?
Somthing like this https://www.redblobgames.com/grids/parts/#hexagon-coordinates

I did find `hex_corners()` function https://docs.rs/hexx/0.13.0/src/hexx/layout.rs.html#101-104
But I can't easily specify a direction to filter only the wanted edge.

Is it something that's planned? I'm willing to contribute if it's not planned
Thanks again 👍

**Comment by ManevilleF** [comment]
That's feasible, basically you would want an API to retrieve a vertex position given a diagonal direction and an edge given a direction?

When you say edge, a `[Vec2; 2]` for example?

About the article you shared there is no plan yet for a real edge/vertex/face concept with relationships but I'm not opposed to it. What kind of use cases would that answer?

**Reply by f-hoenix** [reply]
Hello, thanks for the reply
I ended up doing something like this, it really feels a but hacky but it works so I'll keep this for now
```
#[derive(PartialEq, Eq, Hash)]
struct Edge {
    hex: Hex,
    direction: Direction,
}

impl Edge {
    fn get_edge_coordinates(self: &Self, layout: &HexLayout) -> (Vec2, Vec2) {
        let (a, b) = {
            if layout.orientation == HexOrientation::Pointy {
                match self.direction {
                    Direction::TopRight => {(Direction::Top, Direction::TopLeft)}
                    Direction::Top => {(Direction::TopLeft, Direction::BottomLeft)}
                    Direction::TopLeft => {(Direction::BottomLeft, Direction::Bottom)}
                    Direction::BottomLeft => {(Direction::Bottom, Direction::BottomRight)}
                    Direction::Bottom => {(Direction::BottomRight, Direction::TopRight)}
                    Direction::BottomRight => {(Direction::TopRight, Direction::Top)}
                }
            } else {
                match self.direction {
                    Direction::TopRight => {(Direction::TopRight, Direction::Top)}
                    Direction::Top => {(Direction::Top, Direction::TopLeft)}
                    Direction::TopLeft => {(Direction::TopLeft, Direction::BottomLeft)}
                    Direction::BottomLeft => {(Direction::BottomLeft, Direction::Bottom)}
                    Direction::Bottom => {(Direction::Bottom, Direction::BottomRight)}
                    Direction::BottomRight => {(Direction::BottomRight, Direction::TopRight)}
                }
            }
        };
        let corners = layout.hex_corners(self.hex);
        (corners[a as usize], corners[b as usize])
    }
}
```

>About the article you shared there is no plan yet for a real edge/vertex/face concept with relationships but I'm not opposed to it. What kind of use cases would that answer?

My use case is to draw walls around 'city tiles' (shape of tiles around a city center tile). 
I'll also try to see if I can also use a hybrid `hex + edge` system for roads, I feel like this would reduce the 'gaminess' of manhattan-like roads in hex games

**Reply by ManevilleF** [reply]
I see, I can probably make something like that.
Btw your implementation should be identical no matter the orientation, coordinates don't change in pointy or flat layouts

**Reply by f-hoenix** [reply]
Thanks for the tip I'll look into it, that's testing results however, I'll search where I might have made an error.

**Reply by f-hoenix** [reply]
It really seems like the order of corners returned by `hex_corners()` is different depending on the layout.

On pointy layout, the first corner is the corner between `(0,0);(1,0);(0;1)` The last is the corner between `(0,0);(0;1);(-1,1)`
On flat layout, the first corner is the corner between `(0,0);(1,0);(1;-1)` The last is the corner between `(0,0);(1;0);(0,1)`

Code snippet:
```
fn draw_gizmos(
    gizmos: &mut Gizmos,
    edge: (Vec2, Vec2),
    color: Color
) {
    gizmos.line_2d(
        Vec2::new(edge.0.x, edge.0.y),
        Vec2::new(edge.1.x, edge.1.y),
        color,
    );
}

fn display_hex_corners(
    grid: Res<HexGrid>,
    mut gizmos: Gizmos,
) {
    let mut previous_corner = None;
    for current_corner in grid.layout.hex_corners(Hex::new(0, 0)) {
        if let Some(c) = previous_corner {
            draw_gizmos(&mut gizmos, (c, current_corner), Color::BLUE);
        }
        previous_corner = Some(current_corner);
    }
}
```
Result
![flat](https://github.com/ManevilleF/hexx/assets/18420593/0bec123d-20f7-471b-aa22-8b6cc275fe87)
![pointy](https://github.com/ManevilleF/hexx/assets/18420593/a0de3cdd-0665-47ba-bb4a-b7479ea6e0f4)

(The line not displayed is the line between the first and the last elements of `hex_corners()`)

**Reply by f-hoenix** [reply]
This is the expected empty edge for the pointy layout
![pointy](https://github.com/ManevilleF/hexx/assets/18420593/327a659d-e00c-41b6-ae69-f3d04c912a33)

**Reply by f-hoenix** [reply]
Since the selected edge is off by 2, I would guess that the angle offset is computed in the wrong direction ?

**Reply by ManevilleF** [reply]
The offset is 30 degrees clockwise, I'm not sure what you mean by wrong direction

**Reply by f-hoenix** [reply]
This part feels weird:
On pointy layout, `grid.layout.hex_corners[0]` is the corner between (0,0);(1,0);(0;1) `grid.layout.hex_corners[5]` is the corner between (0,0);(0;1);(-1,1)
On flat layout, `grid.layout.hex_corners[0]` is the corner between (0,0);(1,0);(1;-1) `grid.layout.hex_corners[5]` is the corner between (0,0);(1;0);(0,1)
Since coordinates are not dependent on the layout, `grid.layout.hex_corners[0]` should yield the same corner independently of the layout

This is why the implem of `get_edge_coordinates` must do weird tricks in `HexOrientation::Pointy` layout
```
impl Edge {
    fn get_edge_coordinates(self: &Self, layout: &HexLayout) -> (Vec2, Vec2) {
        let (a, b) = {
            if layout.orientation == HexOrientation::Pointy {
                match self.direction {
                    Direction::TopRight => {(Direction::Top, Direction::TopLeft)}
                    Direction::Top => {(Direction::TopLeft, Direction::BottomLeft)}
                    Direction::TopLeft => {(Direction::BottomLeft, Direction::Bottom)}
                    Direction::BottomLeft => {(Direction::Bottom, Direction::BottomRight)}
                    Direction::Bottom => {(Direction::BottomRight, Direction::TopRight)}
                    Direction::BottomRight => {(Direction::TopRight, Direction::Top)}
                }
            } else {
                match self.direction {
                    Direction::TopRight => {(Direction::TopRight, Direction::Top)}
                    Direction::Top => {(Direction::Top, Direction::TopLeft)}
                    Direction::TopLeft => {(Direction::TopLeft, Direction::BottomLeft)}
                    Direction::BottomLeft => {(Direction::BottomLeft, Direction::Bottom)}
                    Direction::Bottom => {(Direction::Bottom, Direction::BottomRight)}
                    Direction::BottomRight => {(Direction::BottomRight, Direction::TopRight)}
                }
            }
        };
        let corners = layout.hex_corners(self.hex);
        (corners[a as usize], corners[b as usize])
    }
}
```

**Reply by ManevilleF** [reply]
The vertices are always in the same order and between the same coordinates, they follow `Direction::ALL_DIRECTIONS`. There might be an issue, it should probably use `DiagonalDirection` instead, as it represents vertex direction when `Direction` represents edge direction

I have some code cleaning that up and introducing grid types but it's not ready yet

**Comment by ManevilleF** [comment]
@f-hoenix Could you try the draft PR #154 and see if that could fit your use case ?

**Reply by f-hoenix** [reply]
I'll try this, it looks like it fits my use case yes
